snooty/semanticparser.py

```diff
@@ -79,20 +104,77 @@ def toctree(self, pages: Dict[FileId, Page]) -> Dict[str, SerializableType]:
             slug = fileid.without_known_suffix
             fileid_dict[slug] = fileid
 
-        slug_title: Dict[str, Any] = cast(Dict[str, Any], self.slug_title(pages))[
-            "slugToTitle"
-        ]
+        if not self.slug_title:
+            self.build_slug_title(pages)
 
         # Build the toctree
         root: Dict["str", Any] = {}
         ast: Dict[str, Any] = cast(Dict[str, Any], pages[starting_fileid].ast)
-        find_toctree_nodes(starting_fileid, ast, pages, root, fileid_dict, slug_title)
+
+        find_toctree_nodes(
+            starting_fileid,
+            ast,
+            pages,
+            root,
+            fileid_dict,
+            self.slug_title["slugToTitle"],
+        )
 
         toctree["toctree"] = root
         toctree["toctree"]["title"] = self.project_config.name
         toctree["toctree"]["slug"] = "/"
 
-        return toctree
+        self.toctree = toctree
+
+        return self.toctree
+
+    def breadcrumbs(self, pages: Dict[FileId, Page]) -> Dict[str, SerializableType]:
+        page_dict: Dict[str, Any] = {}
+        if not self.toctree:
+            self.build_toctree(pages)
+
+        all_paths: List[Any] = []
+
+        # Find all node to leaf paths for each node in the toctree
+        if children_exist(self.toctree["toctree"]):
+            for node in self.toctree["toctree"]["children"]:
+                paths: List[str] = []
+                get_paths(node, [], paths)
+                all_paths.extend(paths)
+
+        # Populate page_dict with a list of all possible paths for each slug
+        for path in all_paths:
+            path = [self.toctree["toctree"]["slug"]] + path
+            reversed_path = path[::-1]
+            for i in range(len(reversed_path)):
+                slug = remove_leading_slash(path[i])
+                if path[:i]:
+                    if slug not in page_dict:
+                        page_dict[slug] = [path[:i]]
+                    else:
+                        if path[:i] not in page_dict[slug]:
+                            page_dict[slug].append(path[:i])
+
+        self.pages = {"pages": page_dict}
+
+        return self.pages
+
+
+# Helper function used to retrieve the breadcrumbs for a particular slug
+def get_paths(root: Dict[str, Any], path: List[str], all_paths: List[Any]) -> None:
+    if not root:
+        return
+    if not children_exist(root) or children_exist(root) and len(root["children"]) == 0:
```

Comment:
[nit] Parentheses may clarify this a little
```suggestion
    if not children_exist(root) or (children_exist(root) and len(root["children"]) == 0):
```